README.md
<h1 align="center"> StoryDrop - CS50x Final Project </h1> <br>
<p align="center">
  <a href="https://youtu.be/pQeBUAI9mR0">
    <img alt="StoryDrop" title="StoryDrop" src="https://user-images.githubusercontent.com/5073946/46306780-7ce5fd80-c5b5-11e8-8ef9-06d8b095c6d7.png" width="640"/>
  </a>
</p>

<p align="center">
  Make book publishing more simplified and more democratised.<br/>
  Built with <a href="https://www.w3.org/html/">HTML</a>, <a href="https://www.w3.org/Style/CSS/">CSS</a>, <a href="https://ecma-international.org/ecma-262/5.1/">JavaScript</a> and <a href="https://secure.php.net/">PHP</a>.
</p>

<hr>

<p align="center">
  <b>Build Status</b>
  <br />
  <br />
  <a href="https://travis-ci.org/SuriyaaKudoIsc/storydrop">
    <img alt="Travis CI Build Status" title="Travis CI Build Status" src="https://img.shields.io/travis/SuriyaaKudoIsc/storydrop.svg?style=flat-square">
  </a>
  <a href="https://circleci.com/gh/SuriyaaKudoIsc/storydrop">
    <img alt="CircleCI Build Status" title="CircleCI Build Status" src="https://img.shields.io/circleci/project/github/SuriyaaKudoIsc/storydrop.svg?style=flat-square">
  </a>
  <br />
  <br />
  <b>Facts</b>
  <br />
  <br />
  <a href="https://github.com/SuriyaaKudoIsc/storydrop/">
    <img alt="GitHub code size in bytes" title="GitHub code size in bytes" src="https://img.shields.io/github/languages/code-size/badges/shields.svg?style=flat-square">
  </a>
  <a href="https://github.com/SuriyaaKudoIsc/storydrop/graphs/contributors">
    <img alt="GitHub contributors" title="GitHub contributors" src="https://img.shields.io/github/contributors/SuriyaaKudoIsc/storydrop.svg?style=flat-square">
  </a>
  <a href="http://makeapullrequest.com/">
    <img alt="PRs Welcome" title="PRs Welcome" src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square">
  </a>
  <a href="https://github.com/SuriyaaKudoIsc/storydrop/issues">
    <img alt="GitHub issues" title="GitHub issues" src="https://img.shields.io/github/issues-raw/SuriyaaKudoIsc/storydrop.svg?style=flat-square">
  </a>
  <a href="https://github.com/SuriyaaKudoIsc/storydrop/pulls">
    <img alt="GitHub pull requests" title="GitHub pull requests" src="https://img.shields.io/github/issues-pr-raw/SuriyaaKudoIsc/storydrop.svg?style=flat-square">
  </a>
</p>

<hr>

## Table of Contents

- [Introduction](#introduction)
- [Story Behind](#story-behind)
- [Features](#features)
- [Feedback](#feedback)
- [Dependencies](#dependencies)
- [Contributors](#contributors)
- [Development / Installation Process](#development--installation-process)
- [Contributors](#contributors)
- [Supporters, Backers and Sponsors (Acknowledgments)](#supporters-backers-and-sponsors-acknowledgments)
- [Copyright](#copyright)

## Introduction

<p align="center">
  <a href="https://cs50.harvard.edu/">
    <img alt="This is CS50x" title="This is CS50x" src="https://img.shields.io/badge/This%20is-CS50-lightgrey.svg?longCache=true&style=for-the-badge" width="400"/>
  </a>
</p>

**This is StoryDrop.** A revolutionizing web platform for publishing books (with `.pdf` file type) online.

### Why do we need it?
Many manuscripts from authors, writers and creators are often rejected by publishers day by day:

<p align="center">
    <img alt="The Peanuts - Send me 50 thousand dollars" title="The Peanuts - Send me 50 thousand dollars" src="https://i.pinimg.com/736x/ef/76/6e/ef766e9edc191052730349bb71cf0e4d--writer-humor-writing-inspiration.jpg" width="500"/>
  <br/>
  <small>Copyright &copy; Charles M. Schulz / United Feature Syndicate, Inc. All rights reserved.</small>
</p>

The StoryDrop platform allows everyone publishing their **OWN** story and (e-)books effortless and hurdles-free.

### How does it work?
The platform maintainer or the person/team/company who/that is responsible for running the platform review(s) newly uploaded books in the `uploads/new/` folder. After a fast and open verification process the books are going to be published online at `uploads/verified/`.

*No more publishing companies or dastardly editors that are greedy for money!*

### Vision
> **StoryDrop makes book publishing more simplified and more democratised.** *- [Suriyaa R. Sundararuban][suriyaa] (creator of StoryDrop)*

## Story behind
StoryDrop is the final project made by [Suriyaa R. Sundararuban][suriyaa] ([@SuriyaaKudoIsc][github]) - a student at the [Technische Universität München (TUM)][tum] - for [Harvard's][harvard] [CS50x 2018][cs50] course.

[![Twitter Follow](https://img.shields.io/twitter/follow/SuriyaaKudoIsc.svg?style=social&label=Follow)](twitter.com)

## Features
* Custom 404 error page at `http(s)://www.your-storydrop-domain.tld/errors/404.html`
* Landing page at `http(s)://www.your-storydrop-domain.tld/`
* Contact page at `http(s)://www.your-storydrop-domain.tld/contact.php`
* Upload page with HTML form for uploading `.pdf` files at `http(s)://www.your-storydrop-domain.tld/upload.php`
* Showcase site at `http(s)://www.your-storydrop-domain.tld/showcase.php`
* ASCII HTML Art Code in `includes/header.php`

## Feedback

Feel free to send me feedback on [Twitter](https://twitter.com/SuriyaaKudoIsc) or [file an issue](https://github.com/SuriyaaKudoIsc/storydrop/issues/new). Feature requests are always welcome. If you wish to contribute, please take a quick look at the [guidelines](./CONTRIBUTING.md)!

If there's anything you'd like to chat about, please send me an [e-mail][email]!

## Dependencies

* [Uniform Server][uniformserver] ([ZeroXIII][zeroxiii])
* [Dropzone.js][dropzone]
* [QuickGallery][qgallery] ([commit `8b70912`][qgallerycommit])
* [PHP Contact Form][firstwebdesigner]
* [all-contributors-cli][allcontrib] ([v5.4.0][allcontribsemver])

## Development / Installation Process

### For developers

* Download the latest [Uniform Server](uniformserver).
* Extract the installation file.
* Clone this project: `git clone https://github.com/SuriyaaKudoIsc/storydrop.git`
* Put this folder in the `www/` directory of the [Uniform Server](uniformserver).
* Run the `UniController.exe` file in the root directory (on Windows), start the Apache & MySQL servers.
* :tada: It works!
* :pencil: Start contributing...

### For production

* Download the latest [Uniform Server](uniformserver).
* Extract the installation file.
* [Download](https://github.com/SuriyaaKudoIsc/storydrop/archive/master.zip) this project.
* Put the `storydrop` folder in the `www/` directory of the [Uniform Server](uniformserver).
* Run the `UniController.exe` file in the root directory (on Windows) and start the Apache & MySQL servers.
* :smiley: Nice. Go to `http://localhost/`
* :tada: You run your own local StoryDrop platform!

### For repo maintainer(s)

Automate [adding contributor acknowledgements][allcontrib]:

	npm run contributors:add -- SuriyaaKudoIsc doc
	npm run contributors:generate

Add/update contributors (View the Emoji key spec [here][allcontribkey]):

	# Add new contributor <username>, who made a contribution of type <contribution>
	all-contributors add <username> <contribution>
	# Example:
	all-contributors add SuriyaaKudoIsc code,doc

Check for missing contributors:

	all-contributors check

## Contributors

This project follows the [all-contributors](https://github.com/kentcdodds/all-contributors) specification and is brought to you by these [awesome contributors](./CONTRIBUTORS.md).

## Supporters, Backers and Sponsors (Acknowledgments)

Thanks for supporting me:

| [<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/8f/EdX.svg/1280px-EdX.svg.png" width="160px;"/><br /><sub><b>edX.org</b>](https://www.edx.org/)</sub><br /><sub>(MOOC)</sub> | [<img src="https://www.tum.de/typo3conf/ext/in2template/Resources/Public/Images/tum-logo.svg" width="160px;"/><br /><sub><b>Technische Universität München</b></sub>](https://www.tum.de/) | [<img src="https://upload.wikimedia.org/wikipedia/en/2/29/Harvard_shield_wreath.svg" width="120px;"/><br /><sub><b>Harvard University</b></sub>](https://www.harvard.edu/) | [<img src="https://cs.harvard.edu/malan/malan6.jpg" width="140px;"/><br /><sub><b>David J. Malan</b>](https://cs.harvard.edu/malan/)</sub><br /><sub>(Harvard's CS50 Professor)</sub> |
|:-:|:-:|:-:|:-:|
| [Financial Assistance][edxfinance] | For being a [TUM][tumcampus] student | Make [CS50][cs50] available worldwide | Explains very comprehensible |

## Copyright

Copyright 2018-present **[Suriyaa R. Sundararuban][suriyaa]** *(Graduate of [CS50x 2018][cs50])*

[suriyaa]: https://about.suriyaa.tk/
[tum]: https://tum.de/
[tumcampus]: https://campus.tum.de/
[cs50]: https://cs50.harvard.edu/
[harvard]: https://www.harvard.edu/
[edxfinance]: https://courses.edx.org/financial-assistance/
[dropzone]: https://www.dropzonejs.com/
[uniformserver]: https://www.uniformserver.com/
[zeroxiii]: https://sourceforge.net/projects/miniserver/files/Uniform%20Server%20ZeroXIII/13_3_2_ZeroXIII/
[allcontrib]: https://www.npmjs.com/package/all-contributors-cli
[allcontribsemver]: https://github.com/jfmengels/all-contributors-cli/releases/tag/v5.4.0
[allcontribkey]: https://github.com/kentcdodds/all-contributors#emoji-key
[qgallery]: https://github.com/mojeda/QuickGallery
[qgallerycommit]: https://github.com/mojeda/QuickGallery/tree/8b709128ec478b7536b4080fe977e806cf473757
[firstwebdesigner]: https://1stwebdesigner.com/php-contact-form-html/
[email]: mailto:suriyaa.sundararuban@tum.de
[github]: https://github.com/SuriyaaKudoIsc


CHANGELOG.md
## Changelog

**30-sep-2018**
* Add Travis CI
* Add CircleCI
* Add .gitignore
* Add .editorconfig
* Create CHANGELOG.md

**29-sep-2018**
* Works/Tested with [Uniform Server](https://www.uniformserver.com/) (Apache 2, PHP 7.1, MySQL5 included / for portable usage)
* Create `showcase.php`
* Add custom [`404.html` error page](./errors/404.html) for Apache 2.4 Server
* Create `.htaccess`
* Add Node.js all-contributor plugin
* Add GitHub templates in `.github` folder
* Create CONTRIBUTING.md

**28-sep-2018**
* Create `robots.txt`
* Create `index.php`
* Create `upload.php`
* Create `assets` folder
* Create `handler` folder
* Create contact.php
* Create `includes` folder
* Create `uploads` folder

**27-sep-2018**
* Initial commit


upload.php
<?php 
include('config.php');
$pageName = 'upload';
$pageTitle = 'Upload';
include('./includes/header.php'); 

// Store directory separator (DIRECTORY_SEPARATOR) to a simple variable. This is just a personal preference as we hate to type long variable name.
$ds = DIRECTORY_SEPARATOR;
// Declare a variable for destination folder.
$storeFolder = 'uploads/new';

if (!empty($_FILES)) {
// If file is sent to the page, store the file object to a temporary variable.
	$tempFile = $_FILES['file']['tmp_name'];
// Create the absolute path of the destination folder.
	$targetPath = dirname( __FILE__ ) . $ds. $storeFolder . $ds;
// Create the absolute path of the uploaded file destination.
	$targetFile =  $targetPath. $_FILES['file']['name'];
// Move uploaded file to destination.
	move_uploaded_file($tempFile,$targetFile);
}
?>

	<div id="content">
		<h2>Upload form</h2>
		<p>Upload your manuscript (only PDF files supported!):</p>

			<div id="dropzone">
				<form action="upload.php" class="dropzone needsclick dz-clickable" id="demo-upload">
				<div class="dz-message needsclick">
					Drop your manuscript files here or click to upload.
					<br />
					<span class="note needsclick">(NOTE: After uploading you can not remove the files again!)</span>
				</div>
				</form>
			</div>
	</div>

<?php include('./includes/footer.php'); ?>


contact.php
<?php 
include('config.php');
$pageName = 'contact';
$pageTitle = 'Contact';
include('./includes/header.php');
?>

	<form action="handler/mail.php" method="POST">
		<fieldset>
			<legend>Contact Information</legend>
			<p>Name (required)</p> <input type="text" name="name" placeholder="Type your name here...">
			<p>Email (required)</p> <input type="text" name="email" placeholder="Type your e-mail address here...">
		</fieldset>
		<br />
		<fieldset>
		<legend>Other Information</legend>
			<p>Website</p> <input type="text" name="website" placeholder="Type your website here...">
			<br />
		</fieldset>
		<br />
		<fieldset>
			<p>Type (required)</p>
			<select name="type" size="1">
			<option value="update">Feedback</option>
			<option value="change">DMCA Report</option>
			<option value="addition">Report Issue</option>
			<option value="new">Feature Request</option>
			</select>
			<legend>Your Message</legend>
			<p>Message (required)</p><textarea name="message" rows="8" cols="29"  placeholder="Type your text message here..."></textarea><br />
			<input type="submit" value="Send"><input type="reset" value="Clear">
		</fieldset>
	</form>

	<!-- Critical CSS Rendering -->
	<link rel="stylesheet" type="text/css" href="assets/css/mail.css?v=1" />
<?php include('./includes/footer.php'); ?>


showcase.php
<?php 
include('config.php');
$pageName = 'showcase';
$pageTitle = 'Showcase';
include('./includes/header.php');
?>
  <div class="container-fluid">
    <div class="row-fluid">
      <div class="col-md-10 gallery">
		<?php
            if($_GET['gallery'] == 'new'){
				// Path to directories to scan
				$new_directory = "./uploads/new/";

				// Only .pdf files allowed!
				$allowed_types = array('pdf');

				if( $allowed_types )
				{
					// Get all files with a .pdf extension.
					$pdflink = glob($new_directory . "*.pdf");

					// Print each file name
					foreach($pdflink as $link) {
						echo "<a href='".$link."'><img src='assets/img/adobe-pdf-logo.png' width='100' /></a>";
					}
				}
			}
			
			if($_GET['gallery'] == 'verified'){
				// Path to directories to scan
				$new_directory = "./uploads/verified/";

				// Only .pdf files allowed!
				$allowed_types = array('pdf');

				if( $allowed_types )
				{
					// Get all files with a .pdf extension.
					$pdflink = glob($new_directory . "*.pdf");

					// Print each file name
					foreach($pdflink as $link) {
						echo "<a href='".$link."'><img src='assets/img/adobe-pdf-logo.png' width='100' /></a>";
					}
				}
			}
        ?>
      </div>
    </div>
  </div>
</body>
</html>


includes/header.php
<?php
/**
 * @package Core
 */

// Handle any universal redirects, etc

// Setup generic data:
if(!$pageTitle) 		$pageTitle = 'Home'; // TODO: localize
if(!$pageKeywords) 		$pageKeywords = siteKeywords; // TODO: localize
if(!$js)				$js = array();
if(!$jsExternal)		$jsExteral = array();
if(!$css)				$css = array();

// Generic site includes
include('config.php');
?>
<!DOCTYPE html>
<html lang="en">
<!--
████████╗██╗  ██╗██╗███████╗    ██╗███████╗                                 
╚══██╔══╝██║  ██║██║██╔════╝    ██║██╔════╝                                 
   ██║   ███████║██║███████╗    ██║███████╗                                 
   ██║   ██╔══██║██║╚════██║    ██║╚════██║                                 
   ██║   ██║  ██║██║███████║    ██║███████║                                 
   ╚═╝   ╚═╝  ╚═╝╚═╝╚══════╝    ╚═╝╚══════╝                                 
                                                                            
███████╗████████╗ ██████╗ ██████╗ ██╗   ██╗██████╗ ██████╗  ██████╗ ██████╗ 
██╔════╝╚══██╔══╝██╔═══██╗██╔══██╗╚██╗ ██╔╝██╔══██╗██╔══██╗██╔═══██╗██╔══██╗
███████╗   ██║   ██║   ██║██████╔╝ ╚████╔╝ ██║  ██║██████╔╝██║   ██║██████╔╝
╚════██║   ██║   ██║   ██║██╔══██╗  ╚██╔╝  ██║  ██║██╔══██╗██║   ██║██╔═══╝ 
███████║   ██║   ╚██████╔╝██║  ██║   ██║   ██████╔╝██║  ██║╚██████╔╝██║     
╚══════╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝   ╚═╝   ╚═════╝ ╚═╝  ╚═╝ ╚═════╝ ╚═╝     
									A CS50x project
												by Suriyaa R. Sundararuban
-->
<head>
	<meta charset="utf-8" />
    <meta name="keywords" content="<?=$pageKeywords?>" />
    <?=isset($pageDescription)?'<meta name="description" content="'.$pageDescription.'" />':''?>
	<title><?=$pageTitle?>: <?=$siteTitle?></title>
	<!-- DNS Prefetching -->
	<link rel="dns-prefetch" href="//ajax.googleapis.com">
    <!-- Global CSS -->
    <link href="assets/css/style.css" media="all" rel="stylesheet" />
    <link href="assets/css/dropzone.css" media="all" rel="stylesheet" />
	<link rel="stylesheet" type="text/css" href="assets/css/style.css?v=1" />
    <link rel="stylesheet" type="text/css" href="//cdn.jsdelivr.net/fancybox/2.1.5/jquery.fancybox.css" media="screen" />
	<!-- Javascript -->
	<script type="text/javascript" src="//cdn.jsdelivr.net/fancybox/2.1.5/jquery.fancybox.pack.js"></script>
    <script type="text/javascript">
      $(document).ready(function() {
        $("a[rel=gallery]").fancybox({
          'transitionIn'    : 'none',
          'transitionOut'   : 'none',
          'titlePosition'   : 'over',
          'titleFormat'   : function(title, currentArray, currentIndex, currentOpts) {
            return '<span id="fancybox-title-over">Image ' + (currentIndex + 1) + ' / ' + currentArray.length + (title.length ? ' &nbsp; ' + title : '') + '</span>';
          }
        });
      });
    </script>
    <script type="text/javascript" src="assets/js/dropzone.js"></script>
	<script type="text/javascript" src="<?=PROTOCOL?>ajax.googleapis.com/ajax/libs/jquery/1/jquery.min.js"></script>
    <?php
	// JS that is hotlinked from a full server path
	for($i=0;$i<count($jsExternal);$i++){
		echo '<script type="text/javascript" src="'.PROTOCOL.$jsLib[$i].'"></script>';
	}
	// Load page specific dynamic JS
	for($i=0;$i<count($js);$i++){
		echo '<script type="text/javascript" src="/js/'.$js[$i].'.js?v='.CACHEBUST.'"></script>';
	}
	// CSS
	for($i=0;$i<count($css);$i++){
		echo '<link rel="stylesheet" type="text/css" href="/css/'.$css[$i].'.css?v='.CACHEBUST.'" />';
	}
?>
</head>
<body>
    <div id="header" class="center"><?=isset($siteTitle)?'<h1>&#10023; '.$siteTitle.' &#10023;</h1>':''?></div>

	<div class="center">
		<nav>
			<a href="index.php">Home</a> | <a href="upload.php">Upload</a></li> | <a href="showcase.php">Showcase</a> (<a href="showcase.php?gallery=verified">Verfied</a> | <a href="showcase.php?gallery=new">New</a>) | <a href="contact.php">Contact</a>
		</nav>
	</div>

	<div class="center">
		<h2><strong>Make book publishing more simplified and more democratised</strong></h2>
	</div>

	<hr>


index.php
<?php 
include('config.php');
$pageName = 'index';
$pageTitle = 'Home';
include('./includes/header.php');
?>

	<div id="content">
		<p><strong>This is StoryDrop.</strong> A revolutionizing web platform for publishing books (with PDF file type) online.</p>

		<h3>Why do we need it?</h3>
		<p>Many manuscripts from authors, writers and creators are often rejected by publishers day by day:</p>

		<p align="center">
			<img alt="The Peanuts - Send me 50 thousand dollars" title="The Peanuts - Send me 50 thousand dollars" src="https://i.pinimg.com/736x/ef/76/6e/ef766e9edc191052730349bb71cf0e4d--writer-humor-writing-inspiration.jpg" width="500"/>
			<br/>
			<small>Copyright &copy; Charles M. Schulz / United Feature Syndicate, Inc. All rights reserved.</small>
		</p>

		<p>The StoryDrop platform allows everyone publishing their <strong>OWN</strong> story and (e-)books effortless and hurdles-free.</p>

		<h3>How does it work?</h3>
		<ul>
			<li>The platform maintainer review(s) newly uploaded books in the <code>uploads/new/</code> folder.</li>
			<li>After a fast and open verification process the books are going to be published online at <code>uploads/verified/</code>.</li>
		</ul>

		<p><strong>No more publishing companies or dastardly editors that are greedy for money!</strong></p>

		<h3>Vision</h3>
		<blockquote>
			<strong>StoryDrop makes book publishing more simplified and more democratised.</strong>
			<br/>
			<p align="right"><i>- <a href="https://about.suriyaa.tk/">Suriyaa R. Sundararuban</a> (Creator of StoryDrop)</i></p>
		</blockquote>

		<h3>Story behind</h3>
		<p>StoryDrop is the final project made by <a href="https://about.suriyaa.tk/">Suriyaa R. Sundararuban</a> - a student at the <a href="https://www.tum.de/">Technische Universität München (TUM)</a> - for <a href="https://www.harvard.edu/">Harvard's</a> <a href="https://cs50.harvard.edu/">CS50x 2018</a> course.</p>
	</div>

<?php include('./includes/footer.php'); ?>
